File: internal/shadow/core/mirrortree_test.go
package core

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestMirrorTreeCopiesRecursively(t *testing.T) {
	src, dest := t.TempDir(), t.TempDir()
	require.NoError(t, os.MkdirAll(filepath.Join(src, ".github", "workflows"), 0o755))
	require.NoError(t, os.WriteFile(filepath.Join(src, "README.md"), []byte("hi"), 0o644))
	require.NoError(t, os.WriteFile(filepath.Join(src, ".github", "workflows", "ci.yaml"), []byte("on: push"), 0o644))

	require.NoError(t, MirrorTree(src, dest))

	b, err := os.ReadFile(filepath.Join(dest, "README.md"))
	require.NoError(t, err)
	assert.Equal(t, "hi", string(b))
	assert.FileExists(t, filepath.Join(dest, ".github", "workflows", "ci.yaml"))
}

func TestMirrorTreeNeverCopiesGit(t *testing.T) {
	src, dest := t.TempDir(), t.TempDir()
	require.NoError(t, os.MkdirAll(filepath.Join(src, ".git"), 0o755))
	require.NoError(t, os.WriteFile(filepath.Join(src, ".git", "HEAD"), []byte("ref: refs/heads/main"), 0o644))
	require.NoError(t, os.WriteFile(filepath.Join(src, "app.js"), []byte("x"), 0o644))

	require.NoError(t, MirrorTree(src, dest))

	assert.FileExists(t, filepath.Join(dest, "app.js"))
	assert.NoDirExists(t, filepath.Join(dest, ".git"))
}

File: shadow/test/mirrorTree.test.ts
import { describe, it, beforeEach, afterEach } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, mkdirSync, writeFileSync, rmSync, existsSync, readFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { mirrorTree } from '../src/core/mirrorTree.ts';

let src: string;
let dest: string;

beforeEach(() => {
  src = mkdtempSync(join(tmpdir(), 'mt-src-'));
  dest = mkdtempSync(join(tmpdir(), 'mt-dest-'));
});
afterEach(() => {
  rmSync(src, { recursive: true, force: true });
  rmSync(dest, { recursive: true, force: true });
});

describe('mirrorTree', () => {
  it('copies the source tree into dest, recursively', () => {
    mkdirSync(join(src, '.github', 'workflows'), { recursive: true });
    writeFileSync(join(src, 'README.md'), 'hi');
    writeFileSync(join(src, '.github', 'workflows', 'ci.yaml'), 'on: push');

    mirrorTree(src, dest);

    assert.equal(readFileSync(join(dest, 'README.md'), 'utf8'), 'hi');
    assert.ok(existsSync(join(dest, '.github', 'workflows', 'ci.yaml')));
  });

  it('never copies the source .git directory', () => {
    mkdirSync(join(src, '.git'), { recursive: true });
    writeFileSync(join(src, '.git', 'HEAD'), 'ref: refs/heads/main');
    writeFileSync(join(src, 'app.js'), 'x');

    mirrorTree(src, dest);

    assert.ok(existsSync(join(dest, 'app.js')));
    assert.ok(!existsSync(join(dest, '.git')));
  });
});
